File: backend/src/main/java/com/skewwhiffy/auraltester/notation/factory/ClefFactory.java
package com.skewwhiffy.auraltester.notation.factory;

import com.skewwhiffy.auraltester.notation.model.note.AbsoluteNote;
import com.skewwhiffy.auraltester.notation.model.clef.Clef;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@AllArgsConstructor
public class ClefFactory {
    private final NoteFactory noteFactory;

    public Clef getTreble() {
        return new Clef(
                "treble",
                AbsoluteNote.getMiddleC(),
                noteFactory.getAbsoluteNote("a")
        );
    }

    public Clef getAlto() {
        return new Clef("alto", noteFactory.getAbsoluteNote("D,"), noteFactory.getAbsoluteNote("B"));
    }


    public Clef getTenor() {
        return new Clef("tenor", noteFactory.getAbsoluteNote("B,,"), noteFactory.getAbsoluteNote("G"));
    }


    public Clef getBass() {
        return new Clef("bass", noteFactory.getAbsoluteNote("E,,"), AbsoluteNote.getMiddleC());
    }
}

File: backend/src/main/kotlin/com/skewwhiffy/auraltester/notation/factory/ClefFactory.kt
package com.skewwhiffy.auraltester.notation.factory

import com.skewwhiffy.auraltester.controller.ClefType
import com.skewwhiffy.auraltester.notation.model.clef.Clef
import com.skewwhiffy.auraltester.notation.model.note.AbsoluteNote
import org.springframework.stereotype.Service

@Service
class ClefFactory(private val noteFactory: NoteFactory) {
    fun get(clefType: ClefType) = when (clefType) {
        ClefType.TREBLE -> treble
        ClefType.ALTO -> alto
        ClefType.TENOR -> tenor
        ClefType.BASS -> bass
    }

    val treble: Clef
        get() {
            return Clef(
                ClefType.TREBLE,
                "treble",
                AbsoluteNote.middleC,
                noteFactory.getAbsoluteNote("a")
            )
        }

    val alto: Clef
        get() {
            return Clef(
                ClefType.ALTO,
                "alto",
                noteFactory.getAbsoluteNote("D,"),
                noteFactory.getAbsoluteNote("B")
            )
        }

    val tenor: Clef
        get() {
            return Clef(
                ClefType.TENOR,
                "tenor",
                noteFactory.getAbsoluteNote("B,,"),
                noteFactory.getAbsoluteNote("G")
            )
        }

    val bass: Clef
        get() {
            return Clef(
                ClefType.BASS,
                "bass",
                noteFactory.getAbsoluteNote("E,,"),
                AbsoluteNote.middleC
            )
        }
}
